lib/teleterm/clusters/cluster_databases.go

```diff
@@ -121,3 +122,20 @@ func (c *Cluster) ReissueDBCerts(ctx context.Context, user string, db types.Data
 
 	return nil
 }
+
+// GetAllowedDatabaseUsers returns allowed users for the given database based on the role set.
+func (c *Cluster) GetAllowedDatabaseUsers(ctx context.Context, dbURI string) ([]string, error) {
+	roleSet, err := services.FetchRoles(c.status.Roles, c.clusterClient, c.status.Traits)
```

Comment:
The `tsh db ls -v` version of this code [connects to the proxy first](https://github.com/gravitational/teleport/commit/6bd7bf95c37da01cd13a5f0100706412962633df#diff-dae87d4d3799335d85a5fa6ff17cfcbfb30705ae145c9e67ef70286423676a48R54) before calling `services.FetchRoles`.

Do I need to do this here too? The `Cluster` struct I'm operating on here already has references to `ProfileStatus` and `TeleportClient` that I can use here.

https://github.com/gravitational/teleport/blob/6090379bce2c68b209dfe6c64bb58d6cd83c0592/lib/teleterm/clusters/cluster.go#L43-L46

Reply:
The final version, merged to master, operates the same way for verbose and non-verbose modes.

You do need a definition of the role to parse it. From memory, the profiles only contain the names of the roles.

Reply:
@Tener Sorry, I think I might've worded myself incorrectly. I wanted to ask if I need to do call `tc.ConnectToProxy` and `proxy.ConnectToCurrentCluster` too before calling `services.FetchRoles` rather than just passing `client.ProfileStatus` and `client.TeleportClient` that were already created during tsh daemon's lifetime (when we first establish a connection to a cluster).

I wasn't sure if tsh is calling `tc.ConnectToProxy` and `proxy.ConnectToCurrentCluster` just to get the `cluster` object which is then passed to `services.FetchRoles` or if there's some hidden rule like "You should always call `tc.ConnectToProxy` first before doing anything".

But I just looked at `TeleportClient` methods like `GetRole` and it looks like it's going to call `ConnectToProxy` anyway for me, so I think I'm safe here?

There's a couple of places is Teleterm that call `ConnectToProxy`, but I think they do this only so that they can call methods on `ProxyClient` which don't exist on `TeleportClient`, which is not the case here.

Reply:
Ah, I see, there are no hidden rules here (that I'm aware of 😉 ), I believe that using an existing connection is actually preferable (mainly for performance reasons).